Multi get, pluggable hashing and some code cleanup (#1)

blackmad opened the pull request:
This change implements multi-get, pluggable hashing and some (but not complete) code cleanup

pluggable hashing
- removes the failover options which would be hard to make work with pluggable hashing (if this is important to keep, I will figure out a way to add it back) 
- implemented with the idea in mind of being able to use https://www.npmjs.com/package/hashring
- lots of tests added!

code cleanup
- move to tap v14 so tap.only works
- introduce a constants file, change most (but not all) magic numbers in memjs to use named constants
- start using "server" in many places rather than "serv"

I'd really like to move the eslint version forward so we can use const,  let, =>, and array destructuring, but that seemed like it would be better suited to a separate PR

Commit:
Multi get, pluggable hashing and some code cleanup (#1)

Commit:
remove comment to self

Commit:
Merge remote-tracking branch 'upstream/master'

justjake committed:
.d.ts

justjake committed:
confirmed these were actually Buffers

saschat commented:
Hey @blackmad, thanks so much for all the work.

I have only looked at the pluggable hash function so far.

I don't think removing failover is a good idea for a couple of reasons.
* Without failover, you will be dogpiling servers whenever there is an issue... and in the cloud, there is always an issue eventually.
* For MemCachier caches, servers have a consistent view of the cache, so failovers are highly desirable since you will get the same answer from any server.

The good thing is that I don't think it is hard to keep failover at all. For the default function, you can just leave the "use next server" failover. Personally, I would also recommend the same failover for `hashring` as it is common practice. Btw. feel free to add the `hashring` function to choose a server to `memjs` in a future PR as well. I think @alevy has a TODO about adding one in the code anyway. The good thing about having both in `memjs` is that you can reuse the "use next server" failover.

blackmad commented:
Thanks for the review!

I originally had a version with hashring but noticed that the project had
zero external dependencies so thought you might want to keep it that way,
rewriting it to be pluggable ended up with cleaner code so I'm glad I did
it, but awesome can add back in hashring in the future.

re: failover, sounds like a simple change - I'll try to get in back in
today.



On Mon, Mar 29, 2021 at 4:51 AM Sascha ***@***.***> wrote:

> Hey @blackmad <https://github.com/blackmad>, thanks so much for all the
> work.
>
> I have only looked at the pluggable hash function so far.
>
> I don't think removing failover is a good idea for a couple of reasons.
>
>    - Without failover, you will be dogpiling servers whenever there is an
>    issue... and in the cloud, there is always an issue eventually.
>    - For MemCachier caches, servers have a consistent view of the cache,
>    so failovers are highly desirable since you will get the same answer from
>    any server.
>
> The good thing is that I don't think it is hard to keep failover at all.
> For the default function, you can just leave the "use next server"
> failover. Personally, I would also recommend the same failover for
> hashring as it is common practice. Btw. feel free to add the hashring
> function to choose a server to memjs in a future PR as well. I think
> @alevy <https://github.com/alevy> has a TODO about adding one in the code
> anyway. The good thing about having both in memjs is that you can reuse
> the "use next server" failover.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/memcachier/memjs/pull/145#issuecomment-809197945>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADMZMBP3J2QQUWYDXPW2WDTGBETNANCNFSM4Z3LJ37Q>
> .
>


-- 
*David Blackman*
creative technologist & wandering
help me find my purpose <http://purpose.blackmad.com>

justjake committed:
fix file name

Commit:
add version command (#5)

justjake committed:
fix comment to accurately describe as buffer

justjake committed:
Merge pull request #4 from makenotion/jake--types

Add type file with `getMulti`

Commit:
version and versionAll (#6)

* version and versionAll

* PR feedback

Commit:
no-op lint fix commit

justjake committed:
move sources into src/

justjake committed:
tsconfig

justjake committed:
build with tsc

justjake committed:
rename .js to .ts

justjake committed:
constants: convert to module

justjake committed:
tsify header.ts

justjake committed:
protocol.ts

justjake committed:
utils.ts

justjake committed:
serializer.ts

justjake committed:
server.ts

justjake committed:
WIP

justjake committed:
WIP2

justjake committed:
memjs.ts

justjake committed:
fix exports

justjake committed:
revert changes to tests

justjake committed:
fix tests

Commit:
WIP TSify more

Commit:
TSify tests (poorly?)

Commit:
all tests pass!

Commit:
better tap test type

saschat commented:
Any update on re-adding failover?

justjake committed:
Update src/memjs/header.ts

Co-authored-by: David Blackman <blackmad@makenotion.com>

justjake committed:
Update src/memjs/memjs.ts

Co-authored-by: David Blackman <blackmad@makenotion.com>

justjake committed:
Update src/memjs/memjs.ts

Co-authored-by: David Blackman <blackmad@makenotion.com>

justjake committed:
Update src/memjs/memjs.ts

Co-authored-by: David Blackman <blackmad@makenotion.com>

justjake committed:
Update src/memjs/memjs.ts

Co-authored-by: David Blackman <blackmad@makenotion.com>

justjake committed:
ci

justjake committed:
Merge pull request #8 from makenotion/jake--typescript

Convert to typescript

justjake committed:
return cas from get + getMulti

justjake committed:
move test into src

justjake committed:
mess with imports

justjake committed:
bench seems busted...

justjake committed:
ci on every branch pr?

justjake committed:
test includes tsc now

justjake committed:
pull request target??

justjake committed:
fix getmulti result

justjake committed:
newer versions of makeRequestBuffer

justjake committed:
set: pass through CAS token

justjake committed:
remove no-op fill(0), since Buffer.alloc returns zeroed buffer these days

justjake committed:
verify cas is passed through to server

justjake committed:
refactor to remove repeated error handling boilerplate

justjake committed:
correctly handle CAS failure

justjake committed:
cli tester thingy

justjake committed:
make cli self-executing

justjake committed:
remove ambiguous getmulti case

justjake committed:
remove unused logger

justjake committed:
more commento

justjake committed:
test set w/ cas

justjake committed:
Merge pull request #9 from makenotion/jake--cas

Add support for Compare And Swap (CAS) to get, getMulti, and set

justjake committed:
WIP: remove callbacks from public methods (#10)

* WIP: remove callbacks from public methods

* remove some !

* finish with methods that use this.perform

* wip: fix test

* wip 2 more tests

* all tests pass

justjake committed:
Check in and productionize typescript build (#11)

* build settings for node@12

* node>=12

* set up everything to ensure we check in lib

* smoketest

* dont check in tests tho

* ensure-clean: print the diff pls

* ???

justjake committed:
Make getMulti return type non-null (#12)

* getMulti not null

* derf

blackmad commented:
apologies, the failover change is almost done here: https://github.com/makenotion/memjs/pull/7 but we've been making more substantial changes to the library that's taken precedence - converting to typescript, deprecating the callback API, and removing code to support deprecated options APIs.

It's worth a conversation to see if you would be interested in upstreaming this major change at some point or if we should more officially fork it.

saschat commented:
These are more substantial changes indeed ;-)

We are happy to see interest and activity in `memjs` and `memjs` can definitely use some modernization. We are thus open to quite a few options such as
1. Transfer ownership of `memjs`
2. Add you as a collaborator
3. Upstreaming changes without (1) or (2)

What is important to us is that certain features/options such as failover are not removed since there are many people using those. @blackmad what do you think about that?

Commit:
error on some weird edge cases we are having trouble debugging (#13)

Commit:
Fix incr seq timing in get milti (#14)

* error on some weird edge cases we are having trouble debugging

* fix race in getMulti

tarwin commented:
Is the pluggable Hash going to be rolled into the main branch anytime soon? I'd love to be able to use this but need to write our own Hash support at the moment (horrible old PHP `memcache` compatible CRC32 hash urgh).

saschat commented:
@blackmad any update? Have you thought about the 3 options above?

Commit:
added callbacks to versionAll for better errors

Commit:
readme, gitignore, linting

Commit:
naming and notation

Commit:
naming

Commit:
comments

Commit:
didnt compile

arturoolvera committed:
Merge pull request #15 from makenotion/arturo--add-versionAll-callbacks

added callbacks to versionAll

leon-root committed:
Hanlde server restart

leon-root committed:
nit, remove white space

leon-root committed:
Merge pull request #16 from makenotion/leon/socket.close

Auto Reconnect when socket is closed

schleyfox committed:
Add getMultiWithErrors

schleyfox committed:
Test getMultiWithErrors

schleyfox committed:
Merge pull request #17 from makenotion/benjamin-getmultiwitherrors

Add getMultiWithErrors

schleyfox committed:
Error out waiting clients on close, handle too many connections

Previous behavior was to hang forever on unexpected close or unhandled
error response.

schleyfox committed:
Comment error string

schleyfox committed:
Merge pull request #18 from makenotion/benjamin--handle-close-and-too-many-conns

Error out waiting clients on close, handle too many connections

schleyfox committed:
Reset responseBuffer on error, close, and new connections

schleyfox committed:
Merge pull request #19 from makenotion/benjamin--reset-buffers-on-reconnect

Reset responseBuffer on error, close, and new connections

leon-root committed:
bump version to clear circleCI cache

leon-root committed:
Merge pull request #20 from makenotion/bump.version

bump version to clear circleCI cache

schleyfox committed:
Add versionAllWithErrors

schleyfox committed:
add types

schleyfox committed:
Add tests

schleyfox committed:
bump version

schleyfox committed:
Merge pull request #21 from makenotion/benjamin--versionallwitherrors

Add versionAllWithErrors to return info on which hosts succeeded or failed

schleyfox committed:
Use sock.destroy() instead of end() on timeout, 1.2.4-NOTION.1

Unifies error handling/resetting state in error()

schleyfox committed:
Merge pull request #22 from makenotion/benjamin--destroy-socket-on-timeout

Destroy() socket on timeout
